pkg/networking/session.go

```diff
@@ -0,0 +1,429 @@
+		s.logger.Error("Failed to read handshake from connection", "error", err)
+		return err
+	}
+	s.logger.Debug("Handshake successfully read")
+
+	if !s.config.protocol.IsAcceptableHandshake(hs) {
+		s.logger.Error("Handshake is not acceptable")
+		return ErrUnacceptableHandshake
+	}
+	// Handshake is acceptable, we can switch the session into established state.
+	s.established.Store(true)
+	s.config.handler.OnHandshake(s, hs)
+	return nil
+}
+
+func (s *Session) readMessage(hdr Header) error {
+	// Read the header
+	if _, err := hdr.ReadFrom(s.bufRead); err != nil {
+		if errors.Is(err, io.EOF) {
+			return ErrConnectionClosedOnRead
+		}
+		if errMsg := err.Error(); strings.Contains(errMsg, "closed") ||
+			strings.Contains(errMsg, "reset by peer") ||
+			strings.Contains(errMsg, "broken pipe") { // In Docker network built on top of pipe, we get this error on close.
+			return errors.Join(ErrConnectionClosedOnRead, err) // Wrap the error with ErrConnectionClosedOnRead.
+		}
+		s.logger.Error("Failed to read header", "error", err)
+		return err
+	}
+	if !s.config.protocol.IsAcceptableMessage(hdr) {
+		// We have to discard the remaining part of the message.
+		if _, err := io.CopyN(io.Discard, s.bufRead, int64(hdr.PayloadLength())); err != nil {
+			s.logger.Error("Failed to discard message", "error", err)
+			return err
+		}
+	}
+	// Read the new data
+	if err := s.readMessagePayload(hdr, s.bufRead); err != nil {
+		s.logger.Error("Failed to read message", "error", err)
+		return err
+	}
+	return nil
+}
+
+func (s *Session) readMessagePayload(hdr Header, conn io.Reader) error {
+	// Wrap in a limited reader
+	s.logger.Debug("Reading message payload", "len", hdr.PayloadLength())
+	conn = io.LimitReader(conn, int64(hdr.PayloadLength()))
+
+	// Copy into buffer
+	s.receiveLock.Lock()
+	defer s.receiveLock.Unlock()
+
+	if s.receiveBuffer == nil {
+		// Allocate the receiving buffer just-in-time to fit the full message.
+		s.receiveBuffer = bytes.NewBuffer(make([]byte, 0, hdr.HeaderLength()+hdr.PayloadLength()))
+	}
+	defer s.receiveBuffer.Reset()
+	_, err := hdr.WriteTo(s.receiveBuffer)
+	if err != nil {
+		s.logger.Error("Failed to write header to receiving buffer", "error", err)
+		return err
+	}
+	n, err := io.Copy(s.receiveBuffer, conn)
+	if err != nil {
+		s.logger.Error("Failed to copy payload to receiving buffer", "error", err)
+		return err
+	}
+	s.logger.Debug("Message payload successfully read", "len", n)
+
+	// We lock the buffer from modification on the time of invocation of OnReceive handler.
+	// The slice of bytes passed into the handler is only valid for the duration of the handler invocation.
+	// So inside the handler better deserialize message or make a copy of the bytes.
+	if s.logger.Enabled(s.ctx, slog.LevelDebug) {
+		s.logger.Debug("Invoking OnReceive handler", "message",
+			base64.StdEncoding.EncodeToString(s.receiveBuffer.Bytes()))
+	}
+	s.config.handler.OnReceive(s, s.receiveBuffer.Bytes()) // Invoke OnReceive handler.
```

Comment:
I think it's important to note that `OnReceive` handler **MUST NOT** save or send anywhere the bytes internally (because it's a reusable buffer for `Session` instance).
On my mind it's better to have `io.Reader` as an argument in `OnReceive` signature.

Reply:
Yes, good point, replaced with io.Reader.